### ISSUE TITLE
The function unescape does not produce UTF-8 string

### ISSUE THREAD

**Issue by Chimrod** [asemio/SZXX#1]
In an UTF-8 word, we can expect each string being encoded in UTF-8. I have been surprised to see some output from SZXX in mixed Latin-1 and UTF-8. This is caused by differents way to encode the characters in the spreadsheet (sometimes, the UTF-8 characters where written as is, sometimes using XML entities).

I’ve found the cause in `Xml.escape_table`, which only transform the char number into string. For example `&#233;` into `['é']`. But the UTF-8 encoding requires to encode the characters >127 on two characters like `[0, 'é']`.

I have implemented my own function (in standard OCaml, with StdLabels) and thought this could interest you:

```ocaml
let escape_table = function
  | "amp" -> "&"
  | "lt" -> "<"
  | "gt" -> ">"
  | "apos" -> "'"
  | "quot" -> "\""
  | s ->
    ( match String.get s 0 with
    | '#' ->
        let len = String.length s in
        let value = String.sub s ~pos:1 ~len:(len - 1) in
        let num =
          match String.get s 1 with
          | 'x' ->
              (* Hexadecimal number. We rely on the property of int_of_string
                 to interpret `0x__` values as hexadecimal values *)
              int_of_string (String.cat "0" value)
          | '0' .. '9' -> int_of_string value
          | _ -> raise (InvalidEncoding s)
        in
        ( try
            let c = Uchar.of_int num in
            let b = Buffer.create 4 in
            let encoder = Uutf.encoder `UTF_8 (`Buffer b) in
            match Uutf.encode encoder (`Uchar c) with
            | `Ok ->
                ignore @@ Uutf.encode encoder `End;
                Buffer.contents b
            | `Partial -> s
          with
        | e -> raise e )
    | _ -> Printf.sprintf "&%s;" s )

```

I have also managed the hexadecimal characters, like `&#xa;` Maybe using Uutf is not necessary, but I didn’t had time to check more.

**Comment by SGrondin** [comment]
@Chimrod 
Thank you! This is fantastic. I'll look into this and release a fixed version shortly.
I'm assuming that everything else worked well for you?

**Comment by Chimrod** [comment]
You’re welcome :) 

A few days after the initial post, the new OCaml release has introduced UTF-8 functions, and I rewrite the code:


```ocaml
let escape_table = function
  | "amp" -> "&"
  | "lt" -> "<"
  | "gt" -> ">"
  | "apos" -> "'"
  | "quot" -> "\""
  | s ->
    ( match String.get s 0 with
    | '#' ->
        let len = String.length s in
        let value = String.sub s ~pos:1 ~len:(len - 1) in
        let num =
          match String.get s 1 with
          | 'x' ->
              (* Hexadecimal number. We rely on the property of int_of_string
                 to interpret `0x__` values as hexadecimal values *)
              int_of_string (String.cat "0" value)
          | '0' .. '9' -> int_of_string value
          | _ -> raise (Errors.InvalidEncoding s)
        in
        let c = Uchar.of_int num in

        let len = Uchar.utf_8_byte_length c in
        let arr = Bytes.create len in
        ( match Bytes.set_utf_8_uchar arr 0 c with
        | 0 -> raise (Errors.InvalidEncoding s)
        | _ -> Bytes.to_string arr )
    | _ -> Printf.sprintf "&%s;" s )
```

It is up to you if you want to stick with OCaml 4.14 or not.

Everything else worked fined; I just have a (minor) comment about the library dependancies: I think that the Date type used in `SZXX.Xlsx.parse_date` comes from a transitive dependancy, because I didn’t found where the type is declared, neither in the library itself, neither in the libraries mentionned in the dune file.

Thanks for your amazing work !

**Comment by SGrondin** [comment]
@Chimrod  Oh I really like the new Unicode functions! But unfortunately I want to keep supporting 4.08-4.13, so I'll use Uutf. [I've used it before](https://github.com/asemio/unicode/blob/main/src/unicode/unicode.ml).

`Date.t` comes from `core_kernel`, I'll add a note to the documentation.

**Comment by SGrondin** [comment]
@Chimrod I ended up rewriting `unescape` to be [much more performant as well as produce utf8 strings](https://github.com/asemio/SZXX/commit/29b523225b459a5a08c282fe53c1b910e30695d3).

Then I removed the dependency on Uutf by [handwriting the utf8 encoder](https://github.com/asemio/SZXX/commit/4c1482a881b7f4b1c48e42f5d574efcd3f876173) 😎 

It will be released as 2.1.1 shortly

**Comment by Chimrod** [comment]
Thinks a lot! I like your code with array and binary shift (but it should deserve a comment explaining what it does… in six month I won’t be able to understand it again !)

**Comment by SGrondin** [comment]
@Chimrod  it's now available on OPAM as SZXX 2.1.1
I'm adding this table as a comment in the code: https://en.wikipedia.org/wiki/UTF-8#Encoding 😄
